### ISSUE TITLE
three question about network detail

### ISSUE THREAD

**Issue by budui** [EndlessSora/TSIT#5]
Hello, thanks your great paper first. I am trying to reproduce your work.  I have three question while coding.

1. What is the network structure of input layer and output layer, who `we put an input layer at the entrance. The feature channel is adjusted to 64 after the input layer, while the resolution remains unchanged.` and output layer who reduce 64 channels feature to image?

for now, I use conv block:

```python
# input layer
nn.Sequential(
    conv_block(self.use_spectral, in_channels, out_channels, kernel_size=7, stride=1,
                       padding_mode="zeros", padding=3, bias=True),
    nn.InstanceNorm2d(out_channels),
    nn.ReLU(inplace=True)
)
# output layer
nn.Sequential(
    conv_block(use_spectral, base_channels, out_channels, kernel_size=7, padding=3, padding_mode="zeros"),
    nn.Tanh()
)
```

2. In your network, `FAdaIN` is placed at the front of the `BatchNorm`(in `FADE`). While do `arbitrary style transfer` task, `the batch size is 1`,  so that `BatchNorm` have the same behavior as instance norm(Is this correct?),  Didn’t the sytle injected by `FAdaIN` disappear?

related code in my implementation:

```python
# main code in AdaIN
self.norm = nn.InstanceNorm2d(num_features, eps, momentum, affine, track_running_stats)
x = self.gamma * x + self.beta

# code in FADE,
nn.BatchNorm2d(num_features=in_channels, affine=False, track_running_stats=True)
```  

3. Despite the above two questions, I have tried my implementation in a unpaired dataset( 6000 human face to  3400 anime face). and I got this result (during training after ~90000 pairs):
![image](https://user-images.githubusercontent.com/16448529/92327122-0fd51f80-f08a-11ea-8ca4-04aa7a5708f4.png)

`content image` `style image` `generated image` 

*batch size = 1, feature matching loss weight = 1, perceptual loss weight = 1*

All images generated in test dataset. It seems that the generated image have the average style,  but not the style of the input style image. Is this a problem with my implementation, or a problem with the network in paper?

**Comment by budui** [comment]
I did another experiment with another set of parameters: 

*batch size = 1, feature matching loss weight = 10, perceptual loss weight = 10* 

This is the intermediate results(~144,000 pairs)

![image](https://user-images.githubusercontent.com/16448529/92337401-0aa0c080-f0dc-11ea-9b02-de74df191c76.png)

It seems that the translation result is still not related to the style image, only related to the content image. Is this because of the `BatchNorm`(batch size=1) behind `AdaIN`?

**Comment by EndlessSora** [comment]
1. Your task is `multi-modal image synthesis`. Have you tried the trick in pix2pix, not using `model.eval()` during inference? We have mentioned this in appendix _Additional training details._
2. For this BN, we use [Synchronized-BatchNorm-PyTorch](https://github.com/vacancy/Synchronized-BatchNorm-PyTorch). I tried and found the style feature did not disappear. Besides, in FAdaIN, there's an `alpha` blending the style feature and the original feature.

**Comment by budui** [comment]
1. I don't use `model.eval()` during inference as you said. (In fact, I didn't even notice this...)
2. your paper said that `The models are trained on 1 NVIDIA Tesla V100 GPU`, and `Synchronized-BatchNorm-PyTorch` mentioned *Note that, for one-GPU or CPU-only case, this module behaves exactly same as the built-in PyTorch implementation*, so I think using which batch norm may not matters.   My question about `FAdaIN->BatchNorm`  is about network design rather than implementation.
3. What's `alpha` in FAdaIN? my implementation nearly copy from https://github.com/rosinality/style-based-gan-pytorch/blob/b01ffcdcbca6d8bcbc5eb402c5d8180f4921aae4/model.py#L267-L284
![image](https://user-images.githubusercontent.com/16448529/92354583-2f655a00-f115-11ea-906a-cb345f92ad29.png)
4. What is the network structure of input layer and output layer?

**Comment by EndlessSora** [comment]
Hi, do you know that BN will track and keep mean and variance during training? Please refer to [torch.nn.BatchNorm2d](https://pytorch.org/docs/stable/generated/torch.nn.BatchNorm2d.html#torch.nn.BatchNorm2d). For `alpha`, please refer to Eq. 14 and Fig. 7 of the [original paper](https://openaccess.thecvf.com/content_ICCV_2017/papers/Huang_Arbitrary_Style_Transfer_ICCV_2017_paper.pdf). The input and output layers you implement are almost correct and they will not affect the results too much. For other details, please stay tuned for the code release. Thank you very much! :-)

**Comment by budui** [comment]
Thank you for your reply! I am looking forward to your code. 
I have understand what `alpha` is, thanks! 

I  have set `track_running_stats=True` in my code. And in my understanding, the `running_mean` and `running_var` is useless during training. BN tracks such statistics and only uses it instead batch statistics in eval mode.  If we do *not using model.eval() during inference* as you said, such statistics are not used. I am wondering if there is any problem with my understanding?  Thank you very much! :-)

**Comment by EndlessSora** [comment]
Yes. You can dig deeper into it later to see why it works.
